snowflake/src/main/java/org/apache/iceberg/snowflake/entities/SnowflakeIdentifier.java

```diff
@@ -0,0 +1,167 @@
+  }
+
+  public static SnowflakeIdentifier ofSchema(String databaseName, String schemaName) {
+    Preconditions.checkArgument(null != databaseName, "databaseName must be non-null");
+    Preconditions.checkArgument(null != schemaName, "schemaName must be non-null");
+    return new SnowflakeIdentifier(databaseName, schemaName, null);
+  }
+
+  public static SnowflakeIdentifier ofTable(
+      String databaseName, String schemaName, String tableName) {
+    Preconditions.checkArgument(null != databaseName, "databaseName must be non-null");
+    Preconditions.checkArgument(null != schemaName, "schemaName must be non-null");
+    Preconditions.checkArgument(null != tableName, "tableName must be non-null");
+    return new SnowflakeIdentifier(databaseName, schemaName, tableName);
+  }
+
+  /**
+   * If type is TABLE, expect non-null databaseName, schemaName, and tableName. If type is SCHEMA,
+   * expect non-null databaseName and schemaName. If type is DATABASE, expect non-null databaseName.
+   * If type is ROOT, expect all of databaseName, schemaName, and tableName to be null.
+   */
+  public Type getType() {
+    if (null != tableName) {
+      return Type.TABLE;
+    } else if (null != schemaName) {
+      return Type.SCHEMA;
+    } else if (null != databaseName) {
+      return Type.DATABASE;
+    } else {
+      return Type.ROOT;
+    }
+  }
+
+  public String getTableName() {
+    return tableName;
+  }
+
+  public String getDatabaseName() {
+    return databaseName;
+  }
+
+  public String getSchemaName() {
+    return schemaName;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    } else if (!(o instanceof SnowflakeIdentifier)) {
+      return false;
+    }
+
+    SnowflakeIdentifier that = (SnowflakeIdentifier) o;
+    return Objects.equal(this.databaseName, that.databaseName)
+        && Objects.equal(this.schemaName, that.schemaName)
+        && Objects.equal(this.tableName, that.tableName);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(databaseName, schemaName, tableName);
+  }
+
+  /** Returns this identifier as a String suitable for use in a Snowflake IDENTIFIER param. */
+  public String toIdentifierString() {
+    switch (getType()) {
+      case TABLE:
+        return String.format("%s.%s.%s", databaseName, schemaName, tableName);
+      case SCHEMA:
+        return String.format("%s.%s", databaseName, schemaName);
+      case DATABASE:
+        return databaseName;
+      default:
+        return "";
+    }
+  }
+
+  @Override
+  public String toString() {
+    return String.format("%s: '%s'", getType(), toIdentifierString());
+  }
+
+  /**
+   * Expects to handle ResultSets representing fully-qualified Snowflake Schema identifiers,
+   * containing "database_name" and "name" (representing schemaName).
+   */
+  public static ResultSetHandler<List<SnowflakeIdentifier>> createSchemaHandler() {
+    return rs -> {
```

Comment:
Can these handlers be static since they don't change based on context?

Reply:
Done, turned into public static finals defined at top of class instead of method calls.